checker/src/call_visitor.rs

```diff
@@ -1948,4 +2099,39 @@ impl<'call, 'block, 'analysis, 'compilation, 'tcx, E>
         }
         Rc::new("dummy argument".to_string())
     }
+
+    /// Extract the tag kind and the control mask from the generic arg of the function call
+    /// underlying `add_tag!` or `has_tag!`. The generic arg should be a specialization of
+    /// the generic `MiraiTag` type, whose two parameters stand for the tag kind and the
+    /// control mask. The tag kind should implement the `MiraiTagKind` trait, and the control
+    /// mask should have type `u128`.
+    fn extract_tag_kind_and_mask(&mut self) -> (DefId, u128) {
+        if let Some(rustc_gen_args) = self.callee_generic_arguments {
+            checked_assume!(rustc_gen_args.len() == 2);
+            if let GenericArgKind::Type(tag_rustc_type) = rustc_gen_args[1].unpack() {
+                if let TyKind::Adt(_, substs_ref) = &tag_rustc_type.kind {
+                    checked_assume!(substs_ref.len() == 2);
+                    if let (
+                        GenericArgKind::Type(tag_kind_rustc_type),
+                        GenericArgKind::Const(tag_mask_rustc_const),
+                    ) = (substs_ref[0].unpack(), substs_ref[1].unpack())
+                    {
+                        let tag_mask_abs = self
+                            .block_visitor
+                            .visit_constant(None, tag_mask_rustc_const);
+                        if let (
+                            TyKind::Adt(adt_def, _),
+                            Expression::CompileTimeConstant(ConstantDomain::U128(data)),
+                        ) = (&tag_kind_rustc_type.kind, &tag_mask_abs.expression)
+                        {
+                            let tag_kind = adt_def.did;
+                            let tag_mask = *data;
+                            return (tag_kind, tag_mask);
+                        }
+                    }
+                }
+            }
+        }
+        assume_unreachable!();
```

Comment:
Always justify assumptions via a comment. And NEVER assume that user input is correct.